File: examples/node/collect.js
'use strict';

/**
 * Example: collect from a customer, then wait for the transaction to settle.
 *
 * Run:
 *   MARVIN_API_KEY=... node collect.js
 *   # or, with a .env file (Node 20.6+):  node --env-file=.env collect.js
 *
 * Mobile-money collects are asynchronous: the customer gets a USSD/app prompt,
 * so `collect` typically returns PENDING and we poll `waitForCompletion`.
 */

const crypto = require('node:crypto');
const { MarvinPayClient, MarvinPayError } = require('../../sdks/node/marvin-pay.js');

const apiKey = process.env.MARVIN_API_KEY;
if (!apiKey) {
  console.error('Set MARVIN_API_KEY (see .env.example).');
  process.exit(1);
}

const client = new MarvinPayClient({
  apiKey,
  baseUrl: process.env.MARVIN_BASE_URL, // optional; defaults to production
});

async function main() {
  // transaction_id is YOUR unique reference. The SDK also uses it as the
  // idempotency key, so re-running with the same id returns the same result.
  const transactionId = `example-collect-${crypto.randomUUID()}`;

  const paymentRequest = {
    country_code: 'CM', // currency + country always travel together
    currency: 'XAF',
    amount: 5000, // whole number, 100–500000
    mobile_number: '<your-test-msisdn>',
    payment_method: 'mtn_cm', // from client.getPaymentMethods('CM')
    transaction_id: transactionId,
    beneficiary_name: 'Jane Payer',
    description: 'Example collect',
    fee_bearer: 'MERCHANT', // or 'CUSTOMER' to gross up the payer
    // customer_email: 'jane@example.com', // optional → sends a receipt
  };

  console.log(`Submitting collect ${transactionId} ...`);
  const result = await client.collect(paymentRequest);
  console.log('Accepted:', {
    transaction_id: result.transaction_id,
    transaction_status: result.transaction_status,
    message: result.message,
    partner_transaction_id: result.partner_transaction_id,
    replayed: result.idempotencyReplayed === true,
  });

  console.log('Waiting for completion (poll: 5s start, backoff to 60s, 10m cap) ...');
  const final = await client.waitForCompletion(transactionId);
  console.log('Final status:', {
    raw: final.transaction_status,
    normalized: MarvinPayClient.normalizeStatus(final.transaction_status),
  });
}

main().catch((err) => {
  if (err instanceof MarvinPayError) {
    console.error(`MarvinPayError ${err.httpStatus} (${err.code || 'HTTP'}): ${err.message}`);
    if (err.body) console.error('body:', err.body);
  } else {
    console.error(err);
  }
  process.exit(1);
});

File: examples/python/collect.py
"""Collect from a customer, then wait for the outcome.

Runs a mobile-money collect (payer -> merchant) and polls until the transaction
reaches a terminal state.

Config is read from the environment:
    MARVIN_API_KEY      (required)  your merchant API key
    MARVIN_BASE_URL     (optional)  defaults to production
    MARVIN_COUNTRY      (optional)  ISO-3166 alpha-2, default "CM"
    MARVIN_CURRENCY     (optional)  ISO-4217, default "XAF"
    MARVIN_PAYMENT_METHOD (optional) provider name, default "mtn_cm"
    MARVIN_MOBILE_NUMBER (optional) payer number (test number provided by Marvin Pay)
    MARVIN_AMOUNT       (optional)  whole number, default 5000

Usage:
    export MARVIN_API_KEY=...        # or set it in your shell / .env
    python collect.py
"""

import os
import sys
import time

from marvin_pay import MarvinPayClient, MarvinPayError, normalize_status


def main():
    api_key = os.environ.get("MARVIN_API_KEY")
    if not api_key:
        sys.exit("Set MARVIN_API_KEY (see .env.example).")

    base_url = os.environ.get(
        "MARVIN_BASE_URL", "https://api.marvincorporate.co/api"
    )
    client = MarvinPayClient(api_key=api_key, base_url=base_url)

    # Your own unique reference for this transaction.
    transaction_id = f"order-{int(time.time())}"

    payment_request = {
        # Currency and country ALWAYS travel together.
        "country_code": os.environ.get("MARVIN_COUNTRY", "CM"),
        "currency": os.environ.get("MARVIN_CURRENCY", "XAF"),
        # Whole number only (XAF/XOF have no minor units), range 100..500000.
        "amount": int(os.environ.get("MARVIN_AMOUNT", "5000")),
        "mobile_number": os.environ.get("MARVIN_MOBILE_NUMBER", "<your-test-msisdn>"),
        "payment_method": os.environ.get("MARVIN_PAYMENT_METHOD", "mtn_cm"),
        "transaction_id": transaction_id,
        "description": "SDK example collect",
        # "customer_email": "payer@example.com",  # optional: sends a receipt
        # "fee_bearer": "CUSTOMER",               # optional: MERCHANT (default)
    }

    try:
        # X-Idempotency-Key defaults to the transaction_id.
        result = client.collect(payment_request)
    except MarvinPayError as exc:
        sys.exit(f"collect failed: HTTP {exc.status_code} {exc.message}")

    print("Collect initiated:")
    print(f"  transaction_id     : {result.get('transaction_id')}")
    print(f"  message            : {result.get('message')}")
    print(f"  transaction_status : {result.get('transaction_status')}")
    print("  (a 200/202 does NOT mean money moved — confirming below)")

    if normalize_status(result.get("transaction_status")) in ("SUCCEEDED", "FAILED"):
        return

    print("\nWaiting for a terminal status (this can take a minute)...")
    try:
        final = client.wait_for_completion(transaction_id)
    except MarvinPayError as exc:
        sys.exit(f"still pending: {exc.message}")

    print(f"Final status: {normalize_status(final.get('transaction_status'))}")
    print(f"Raw: {final}")


if __name__ == "__main__":
    main()
